### PR TITLE
feat: strong typing for `onEnter`, `onStay` and `onLeave`

### PR DESCRIPTION
**Pull request by schiller-manuel**

### CODE REVIEW

**Comment by schiller-manuel** [comment]
Performance before this change:

```
Files:                         589
Lines of Library:            38118
Lines of Definitions:        77155
Lines of TypeScript:         12777
Lines of JavaScript:             0
Lines of JSON:                   0
Lines of Other:                  0
Identifiers:                113524
Symbols:                    190002
Types:                       50723
Instantiations:             564441
Memory used:               230872K
Assignability cache size:    34134
Identity cache size:          3717
Subtype cache size:              0
Strict subtype cache size:       0
I/O Read time:               0.02s
Parse time:                  0.24s
ResolveModule time:          0.03s
ResolveTypeReference time:   0.00s
ResolveLibrary time:         0.01s
Program time:                0.33s
Bind time:                   0.11s
Check time:                  2.17s
printTime time:              0.00s
Emit time:                   0.00s
Total time:                  2.61s
```

Performance after this change:

```
Files:                         589
Lines of Library:            38118
Lines of Definitions:        77155
Lines of TypeScript:         12777
Lines of JavaScript:             0
Lines of JSON:                   0
Lines of Other:                  0
Identifiers:                113608
Symbols:                    199432
Types:                       51680
Instantiations:             570731
Memory used:               244387K
Assignability cache size:    34191
Identity cache size:          3717
Subtype cache size:              0
Strict subtype cache size:       0
I/O Read time:               0.02s
Parse time:                  0.25s
ResolveModule time:          0.03s
ResolveTypeReference time:   0.00s
ResolveLibrary time:         0.01s
Program time:                0.35s
Bind time:                   0.12s
Check time:                  2.32s
printTime time:              0.00s
Emit time:                   0.00s
Total time:                  2.79s
```

=> Instantiations increased from 564441 to 570731

File: packages/react-router/src/Matches.tsx [hunk 1]
```diff
@@ -84,7 +69,35 @@ export interface RouteMatch<
   minPendingPromise?: ControlledPromise<void>
 }
 
-export type AnyRouteMatch = RouteMatch<any, any>
+export type MakeRouteMatch<
+  TRouteTree extends AnyRoute = RegisteredRouter['routeTree'],
+  TRouteId extends RouteIds<TRouteTree> = ParseRoute<TRouteTree>['id'],
+  TReturnIntersection extends boolean = false,
+  TTypes extends RouteById<TRouteTree, TRouteId>['types'] = RouteById<
```
Comment (chorobin) [inline packages/react-router/src/Matches.tsx:76]: Can't this constraint be more loose? `AnyRoute['types']`?

File: packages/react-router/src/Matches.tsx [hunk 2]
```diff
@@ -84,7 +69,35 @@ export interface RouteMatch<
...
+    TRouteTree,
+    TRouteId
+  >['types'],
+  TAllParams = TReturnIntersection extends false
+    ? TTypes['allParams']
+    : Expand<Partial<AllParams<TRouteTree>>>,
+  TFullSearchSchema = TReturnIntersection extends false
+    ? TTypes['fullSearchSchema']
+    : Expand<Partial<FullSearchSchema<TRouteTree>>>,
```
Comment (chorobin) [inline packages/react-router/src/Matches.tsx:85]: Do we need to `Expand` here? I usually prefer to do it when the user actually needs the prettified type so it's done lazily

File: packages/react-router/src/router.ts [hunk 3]
```diff
@@ -162,13 +167,16 @@ export interface RouterErrorSerializer<TSerializedError> {
   deserialize: (err: TSerializedError) => unknown
 }
 
-export interface RouterState<TRouteTree extends AnyRoute = AnyRoute> {
+export interface RouterState<
+  TRouteTree extends AnyRoute = AnyRoute,
+  TRouteMatch extends MakeRouteMatch<TRouteTree> = MakeRouteMatch<TRouteTree>,
```
Comment (chorobin) [inline packages/react-router/src/router.ts:172]: Can the constraint be looser or no constraint at all?  `AnyRouteMatch`?

**Comment by chorobin** [comment]
Can we try to add some type tests for this feature ;) Maybe in s separate pr but it would be good to get a habit of adding to them

**Review by chorobin — APPROVED** [review]

**Pull request merged**